pypesto/profile/profile.py

```diff
@@ -0,0 +1,254 @@
+
+
+class ProfileOptions(dict):
+    """
+    Options for optimization based profiling.
+
+    Parameters
+    ----------
+
+    next_profile_startpoint: function handle, optional
+        function which creates the next startpoint for optimization
+        from the profile from the profile history
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.default_step_size = 0.01
+        self.ratio_min = 0.145
+
+    def __getattr__(self, key):
+        try:
+            return self[key]
+        except KeyError:
+            raise AttributeError(key)
+
+    __setattr__ = dict.__setitem__
+    __delattr__ = dict.__delitem__
+
+    @staticmethod
+    def assert_instance(maybe_options):
+        """
+        Returns a valid options object.
+
+        Parameters
+        ----------
+
+        maybe_options: OptimizeOptions or dict
+        """
+        if isinstance(maybe_options, ProfileOptions):
+            return maybe_options
+        options = ProfileOptions(**maybe_options)
+        return options
+
+
+def profile(
+        problem,
+        result,
+        optimizer,
+        profile_index=None,
+        result_index=1,
+        create_profile_startpoint=None,
+        profile_options=None,
+        optimize_options=None) -> Result:
+    """
+    This is the main function to call to do parameter profiling.
+
+    Parameters
+    ----------
+
+    problem: pypesto.Problem
+        The problem to be solved.
+
+    result: pypesto.Result
+        A result object to initialize profiling and to append the profiling
+        results to. For example, one might append more profiling runs to a
+        previous profile, in order to merge these.
+        The existence of an optimization result is obligatory.
+
+    optimizer: pypesto.Optimizer
+        The optimizer to be used along each profile.
+
+    profile_index: integer array, optional
+        array with parameter indices, whether a profile should
+        be computed (1) or not (0)
+
+    result_index: integer, optional
+        index, from which optimization result profiling should be started
+        (default: global optimum, i.e., index = 1)
+
+    create_profile_startpoint: callable, optional
+        Method for how to create the next start point for profile optimization.
+
+    profile_options: pypesto.ProfileOptions, optional
+        Various options applied to the profile optimization.
+
+    optimize_options: pypesto.OptimizeOptions, optional
+        Various options applied to the optimizer.
+    """
+
+    # profiling indices
+    if profile_index is None:
+        profile_index = np.ones(problem.dim_full)
+
+    # profile startpoint method
+    if create_profile_startpoint is None:
+        def create_next_startpoint(x, par_index, par_direction):
+            return fixed_step(x, par_index, par_direction,
+                              step_size=profile_options.default_step_size)
+
+    # check profiling options
+    if profile_options is None:
```

Comment:
profile_options needs to be instantiated before the previous block

Reply:
Done